POST, PATCH, DELETE with ?select= and empty body or return=minimal throw

wolfgangwalther opened the pull request:
Migrating from 6.0.2 to 7.0.0 I encountered the following:

Because the default for DELETE requests is now `return=minimal`, some of my requests without return-header but with `?select=` in the query are now failing with:

```JSON
{
  "hint": null,
  "details":null,
  "code": "42703",
  "message": "column pg_source.selected_column does not exist"
}
```

Instead of throwing this should just return nothing, as expected.

Digging deeper, the same is happening on PATCH and POST calls with return=minimal. I also received this error when trying to PATCH with an empty-body - and using select at the same time. All this worked before the update to 7.0.0.

Some of this is related to #1393, because `RETURNING 1` doesn't return the columns anymore, which `?select=` tries to access.

All covered by the test cases so far. I added some additional test cases about the payload while debugging - those are already passing, but I figured to keep them in. Why delete test cases, eeh? ;)

steve-chavez commented:
You're right. I can also reproduce the issue. Thanks for catching this up!

More test cases are always good :)

steve-chavez commented on test/Feature/DeleteSpec.hs:
```diff
@@ -27,15 +27,25 @@ spec =
           , matchHeaders = ["Content-Range" <:> "*/1"]
           }
+
+      it "ignores ?select= when return=minimal" $
+        request methodDelete "/items?id=eq.3&select=id" [("Prefer", "return=minimal")] ""
```
Since now we default to the `return=minimal` behavior, I think these tests shouldn't contain the `Prefer: return=minimal`. Having the header confused me a little bit when reading the tests, I think the header should only be used when testing POST. Also, maybe the test should be called like `ignores ?select= when no return=representation`. Others could follow the same kind of naming.

What do you think?

wolfgangwalther replied:
I tend to approach it the other way around. There were some tests (I think those with reduced privileges on columns), where a certain default setting for `return` was expected. This confused me! ;)

I think in general it would be best if all tests were specific about `return=representation` or `return=minimal` and would **never** assume a default. Because then those tests would all still work when the default should change at any point.

There should maybe be 1 test to actually test the default itself, to make sure that this doesn't change unintendedly.

steve-chavez replied:
I see what you mean, that makes sense. Adding a test for the default would be great.

steve-chavez commented on src/PostgREST/Statements.hs:
```diff
@@ -72,6 +72,11 @@ createWriteStatement selectQuery mutateQuery wantSingle isInsert asCsv rep pKeys
 
+  selectF
+    -- prevent using any of the column names in ?select= when no response is returned from the CTE
+    | rep `elem` [None, HeadersOnly] = "SELECT * FROM " <> sourceCTEName
```
Hmm.. here I think a `SELECT null` would suffice for preventing the error. With this you wouldn't need the change made to the `QueryBuilder` I think.

wolfgangwalther replied:
I think for the HeadersOnly case you still need the columns at this point to put together the location string or am I missing something right now?

And that makes me wonder: Is it actually guaranteed that the PK columns are always in the `RETURNING` part?

wolfgangwalther replied:
Like when I do a `POST return=rep select=non_pk_only`.

I remember the foreign keys for the embedded columns get added to the `RETURNING`, but I am not sure about the PKs.

steve-chavez replied:
> Is it actually guaranteed that the PK columns are always in the RETURNING part?

No, but maybe that should be a different fix, check https://github.com/PostgREST/postgrest/issues/1461.

wolfgangwalther replied:
added some test cases for that at #1475 
it's not throwing, but just not returning a location header at all it seems. So yes, this is a problem.

wolfgangwalther commented on src/PostgREST/QueryBuilder.hs:
```diff
@@ -94,7 +94,9 @@ mutateRequestToQuery (Insert mainQi iCols onConflct putConditions returnings) =
-    then "WITH " <> ignoredBody <> "SELECT null WHERE false" -- if there are no columns we cannot do UPDATE table SET {empty}, it'd be invalid syntax
+    -- if there are no columns we cannot do UPDATE table SET {empty}, it'd be invalid syntax
+    -- selecting an empty resultset from mainQi gives us the column names to prevent errors when using &select=
+    then "WITH " <> ignoredBody <> "SELECT * FROM " <> fromQi mainQi <> " WHERE false"
```
> [...] With this you wouldn't need the change made to the `QueryBuilder` I think.

Regarding the `QueryBuilder`: We do need that change here anyways for the `PATCH return=rep empty-body` case.

wolfgangwalther commented:
Improved the test cases as suggested and also realized that with the empty-body PATCH with ?select= overloaded computed columns would throw again. Added a test for that as well and fixed it in `QueryBuilder`.

This change also has the `SELECT NULL` again for the default case, ie if no ?select= has been used.

steve-chavez commented on src/PostgREST/QueryBuilder.hs:
```diff
@@ -105,6 +108,10 @@ mutateRequestToQuery (Update mainQi uCols logicForest returnings) =
         ]
   where
     cols = intercalate ", " (pgFmtIdent <> const " = _." <> pgFmtIdent <$> S.toList uCols)
+    empty_body_returned_columns :: SqlFragment
+    empty_body_returned_columns
+      | null returnings = "NULL"
+      | otherwise       = intercalate ", " (pgFmtColumn (QualifiedIdentifier mempty $ qiName mainQi) <$> returnings)
```
This fragment is looking similar to `returned_columns` in `requestToCallProcQuery`. Do you think that could be factored in a function inside [QueryFragment.hs](https://github.com/PostgREST/postgrest/blob/master/src/PostgREST/Private/QueryFragment.hs)?

steve-chavez replied:
Everything else is looking good. I was seeing if there could be some other way to do the fix but your solution is the simplest one.

steve-chavez replied:
Can be done later. Thanks for the fix!
